Date: July 2, 2017 at 11:08 AM
Topic: identity by state and introgression in wild barley

Hi Connor,

**Chaochih**,**Li**, and I have been working on a project idea that they would like to hand over to one of the undergrads in the lab. This is a relatively simple idea, but it is publishable, fun, and with some diligent effort, could be completed quite quickly.

In DNA resequencing of the exome (coding portion of the genome) of wild barley plants, various research groups have repeatedly found that some wild plants are not completely wild. That is, they show clear evidence of having recently undergone introgression from cultivated barley. So the question is; what parts of the genome can be introgressed from the domesticate and still result in the appearance of a wild barley plant? Are there genomic regions especially likely to be subject to or immune to introgression? Or we could ask simply, what makes a wild barley wild?

Our idea for this study is that we use identity by state (IBS) comparisons of wild barley exome data to cultivated barley to ask if there are genomic regions, across wild barley accessions subject to introgression, that are more commonly shared with cultivated lines?

To actually carry out the project, you will need to familiarize yourself with:
* IBS analysis and introgression - start with this [paper]([http://www.g3journal.org/content/3/11/1945](http://www.g3journal.org/content/3/11/1945)) from our lab
* variant call format (VCF) [files]([http://www.internationalgenome.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-40/](http://www.internationalgenome.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-40/))
* the [PLINK]([https://www.cog-genomics.org/plink2](https://www.cog-genomics.org/plink2)) tool that does IBS analysis and conversion of VCF files to PLINK format
* [IBS]([https://www.cog-genomics.org/plink/1.9/ibd](https://www.cog-genomics.org/plink/1.9/ibd)) (also called identity by descent) analysis in PLINK
* measures of [heterozygosity]([https://www.cog-genomics.org/plink/1.9/basic_stats#ibc](https://www.cog-genomics.org/plink/1.9/basic_stats#ibc)); they could be higher in introgressed regions
* measure of allele frequency [differentiation]([https://www.cog-genomics.org/plink/1.9/basic_stats#fst](https://www.cog-genomics.org/plink/1.9/basic_stats#fst)); they could be lower in regions of crop/wild introgression
* the typical nature of crop-wild introgression, as in this [paper]([http://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1003477](http://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1003477))

Because of some other research in the community that might impinge on our ability to publish this work, weâ€™d like to complete it quickly. So, this is a chance to show some drive, initiative, and entrepreneurship! **Li**, **Chaochih**, and I can all help, but the reading list above should provide a strong point of departure.

Let me know if you have any questions.

Peter

Peter L. Morrell
[http://z.umn.edu/morrell](http://z.umn.edu/morrell)
[http://github.com/pmorrell](http://github.com/pmorrell)
[http://github.com/MorrellLab](http://github.com/MorrellLab)
Dept. of Agronomy & Plant Genetics
University of Minnesota
